src/vivarium/framework/resource.py

```diff
@@ -83,21 +87,28 @@ def add_resources(self, resource_type: str, resource_names: List[str],
                 raise ResourceError  # Component has more than one producer for resource type ...
             self.producer_components.add(producer.__self__.name)
 
-        producer = ResourceProducer(resource_type, resource_names, producer, dependencies)
-        for resource_name in resource_names:
-            key = f'{resource_type}.{resource_name}'
+        producer = self.get_producer(resource_type, resource_names, producer, dependencies)
+        for resource_name in producer.resource_names:
+            key = f'{producer.resource_type}.{resource_name}'
             if key in self.resources:
                 raise ResourceError  # More than one producer for resource ...
             self.resources[key] = producer
 
+    def get_producer(self, resource_type, resource_names, producer, dependencies):
```

Comment:
This should probably be private.  Like lookup tables/pop views/pipelines, this object get's handled directly by other components.